ciongke/cmd/hook/main.go

```diff
@@ -0,0 +1,352 @@
+	} else {
+		githubClient = github.NewClient(tc)
+	}
+
+	config, err := restclient.InClusterConfig()
+	if err != nil {
+		log.Fatalf("Error getting rest client: %s", err)
+	}
+	kubeClient, err := client.New(config)
+	if err != nil {
+		log.Fatalf("Error getting client: %s", err)
+	}
+	kubeBatchClient, err := client.NewBatch(config)
+	if err != nil {
+		log.Fatalf("Error getting batch client: %s", err)
+	}
+
+	s := &Server{
+		Events: make(chan Event),
+
+		Port:         *port,
+		Org:          *org,
+		TeamID:       *teamID,
+		TeamName:     *teamName,
+		GitHubClient: githubClient,
+		HMACSecret:   webhookSecret,
+		DryRun:       *dryRun,
+
+		SourceBucket:  *sourceBucket,
+		ResultsBucket: *jobBucket,
+		TestJobImage:  *jobImage,
+
+		KubeClient:      kubeClient,
+		KubeBatchClient: kubeBatchClient,
+		Namespace:       *namespace,
+	}
+	go func() {
+		for event := range s.Events {
+			go s.handleEvent(event)
+		}
+	}()
+	if err := http.ListenAndServe(":"+strconv.Itoa(s.Port), s); err != nil {
+		log.Fatalf("ListenAndServe returned error: %s", err)
+	}
+}
+
+// ServeHTTP validates an incoming webhook and puts it into the event channel.
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+
+	// Header checks: It must be a POST with an event type and a signature.
+	if r.Method != http.MethodPost {
+		http.Error(w, "405 Method not allowed", http.StatusMethodNotAllowed)
+		return
+	}
+	eventType := r.Header.Get("X-GitHub-Event")
+	if eventType == "" {
+		http.Error(w, "400 Bad Request: Missing X-GitHub-Event Header", http.StatusBadRequest)
+		return
+	}
+	sig := r.Header.Get("X-Hub-Signature")
+	if sig == "" {
+		http.Error(w, "403 Forbidden: Missing X-Hub-Signature", http.StatusForbidden)
+		return
+	}
+
+	// Validate the payload with our HMAC secret.
+	payload, err := github.ValidatePayload(r, s.HMACSecret)
+	if err != nil {
+		http.Error(w, "403 Forbidden: Invalid X-Hub-Signature", http.StatusForbidden)
+		return
+	}
+	fmt.Fprintf(w, "Event received. Have a nice day.")
+	go func() {
+		s.Events <- Event{eventType, payload}
+	}()
+}
+
+// handleEvent unmarshals the payload and dispatches to the correct handler.
+func (s *Server) handleEvent(e Event) {
+	// TODO: "test this" comment
+	switch e.Type {
+	case "pull_request":
+		var pr github.PullRequestEvent
+		if err := json.Unmarshal(e.Payload, &pr); err != nil {
+			log.Printf("Could not unmarshal pull request event payload: %s", err)
+			return
+		}
+		if err := s.handlePullRequestEvent(pr); err != nil {
+			log.Printf("Error handling pull request: %s", err)
+			return
+		}
+	}
+}
+
+// handlePullRequestEvent decides what to do with PullRequestEvents.
+func (s *Server) handlePullRequestEvent(pr github.PullRequestEvent) error {
+	switch pr.Action {
+	case "opened", "reopened", "synchronize":
+		valid, err := s.validAuthor(pr.PullRequest.User.Login)
+		if err != nil {
+			return fmt.Errorf("could not validate author: %s", err)
+		}
+		if valid {
```

Comment:
How do we handle the a PR getting merged?

Will a `synchronize` event fire and what will/should we do?